### ISSUE TITLE
global: use webpack dynamic imports to customize search pages

### ISSUE THREAD

**Issue by mvidalgarcia** [inveniosoftware/invenio-app-rdm#234]
https://webpack.js.org/api/module-methods/#dynamic-expressions-in-import

**Comment by mvidalgarcia** [comment]
Achieved a _hardcoded_ solution of dynamic imports in [`search_app_customizations.js`](https://github.com/inveniosoftware/invenio-app-rdm/blob/master/invenio_app_rdm/theme/assets/semantic-ui/js/invenio_app_rdm/search_app_customizations.js)

### Steps to reproduce

1. [Install invenio-rdm](https://inveniordm.docs.cern.ch/develop/install_locally/)
2. Install `invenio-app-rdm` locally in editable mode.
3. **my-site:** Create a `ResultsList.item.jsx` file in my-site's `/assets/templates/search/` (TBD if this is the best place)
```jsx
import React from "react";
import { Item } from "semantic-ui-react";
import _ from "lodash";
import _truncate from "lodash/truncate";

const RDMRecordResultsListItem = ({ result, index }) => {
  const description = _.get(
    result,
    "metadata.descriptions[0].description",
    "No description"
  );
  return (
    <Item key={index} href={`/records/${result.id}`}>
      <Item.Content>
        <Item.Header>Foo bar</Item.Header>
        <Item.Description>
          My custom description
          {_truncate(description, { length: 200 })}
        </Item.Description>
      </Item.Content>
    </Item>
  );
};

export default RDMRecordResultsListItem
```

4. **invenio-app-rdm**: Implement _hardcoded_ dynamic import mechanism in [`search_app_customizations.js`](https://github.com/inveniosoftware/invenio-app-rdm/blob/master/invenio_app_rdm/theme/assets/semantic-ui/js/invenio_app_rdm/search_app_customizations.js)

```jsx
...

// overrideStore.add("ResultsList.item", RDMRecordResultsListItem);

const component = "ResultsList.item";
import(
  /* webpackMode: "eager" */
  `../../templates/search/${component}.jsx`
)
  .then((module) => {
    overrideStore.add("ResultsList.item", module.default);
  })
  .catch((error) => {
    overrideStore.add("ResultsList.item", RDMRecordResultsListItem);
  });
...
```
We need `webpackMode: "eager"`, otherwise we store an empty element in the `overrideStore` as it's not resolvable.

5. `invenio-cli update && invenio-cli run`

![image](https://user-images.githubusercontent.com/8863238/90513306-b4d19c00-e15f-11ea-977e-a347b4e5be68.png)

**Comment by mvidalgarcia** [comment]
A most elegant solution applying to all component customizations.

```diff
diff --git a/invenio_app_rdm/theme/assets/semantic-ui/js/invenio_app_rdm/search_app_customizations.js b/invenio_app_rdm/theme/assets/semantic-ui/js/invenio_app_rdm/search_app_customizations.js
index ac40761..6187ceb 100644
--- a/invenio_app_rdm/theme/assets/semantic-ui/js/invenio_app_rdm/search_app_customizations.js
+++ b/invenio_app_rdm/theme/assets/semantic-ui/js/invenio_app_rdm/search_app_customizations.js
@@ -11,6 +11,27 @@ import { Card, Input, Item } from "semantic-ui-react";
 import _ from "lodash";
 import _truncate from "lodash/truncate";
 
+async function registerComponent(componentId, defaultComponent) {
+  let component;
+  try {
+    await import(
+      /* webpackMode: "eager" */
+      `../../templates/search/${componentId}.jsx`
+    )
+      .then((module) => {
+        component = module.default;
+      })
+      .catch((error) => {
+        component = defaultComponent;
+      })
+      .finally(() => {
+        overrideStore.add(componentId, component);
+      });
+  } catch (error) {
+    console.log("Catch error: ", error);
+  }
+}
+
 const RDMRecordResultsListItem = ({ result, index }) => {
   const description = _.get(
     result,
@@ -30,7 +51,7 @@ const RDMRecordResultsListItem = ({ result, index }) => {
   );
 };
 
-overrideStore.add("ResultsList.item", RDMRecordResultsListItem);
+registerComponent("ResultsList.item", RDMRecordResultsListItem);
 
 const RDMRecordResultsGridItem = ({ result, index }) => {
   const description = _.get(
@@ -50,7 +71,7 @@ const RDMRecordResultsGridItem = ({ result, index }) => {
   );
 };
 
-overrideStore.add("ResultsGrid.item", RDMRecordResultsGridItem);
+registerComponent("ResultsGrid.item", RDMRecordResultsGridItem);
 
 const RDMRecordSearchBarElement = ({
   placeholder: passedPlaceholder,
@@ -84,4 +105,4 @@ const RDMRecordSearchBarElement = ({
   );
 };
 
-overrideStore.add("SearchBar.element", RDMRecordSearchBarElement);
+registerComponent("SearchBar.element", RDMRecordSearchBarElement);

```

**Comment by slint** [comment]
Closed via #260
